### PR TITLE
ptls modules refactoring

### PR DESCRIPTION
**Pull request by PvtKaefsky**
**Key changes:**
- **Parallelization of preprocessing using the dask library**
The preprocessing time on coles_emb demo was reduced from ~40 seconds to ~25 seconds. CPU: Ryzen 5600x.
[Profiling](https://i.imgur.com/RRMIqN9.png)

- **Refactoring the preprocessing module**
The preprocessing time on the coles_emb demo was reduced from ~25 seconds to ~10 seconds. CPU: Ryzen 5600x.
[Profiling](https://i.imgur.com/EXe0VVT.png)

- **Refactoring of data loading modules**
Improved code readability, optimized default parameters experimentally on alphabattle and gender scenarios (GPU: A6000, 3080), fixed bugs.

- **Encoder refactoring** 
Improved code readability, compatibility with new changes, optimization of some functions.

**Result:** 
Reduced preprocessing time from 40 to 10 seconds. Improved code readability, updated typehints, optimized default parameters.

### CODE REVIEW

File: ptls/data_load/datasets/augmentation_dataset.py
Comment (FuTSy13) [inline ptls/data_load/datasets/augmentation_dataset.py:1]: Из ассерта убрали поясняющее сообщение
Предлагаю заменить на какой-то более информативный Exception, например TypeError

File: ptls/data_load/feature_dict.py
Comment (FuTSy13) [inline ptls/data_load/feature_dict.py:1]: is_seq_feature используется не только в этом файле, и изменение сигнатуры функции может сломать код в неожиданных местах
В частности вижу .is_seq_feature (k,v ) в ptls/data_load/iterable_processing/feature_filter.py

File: ptls/data_load/iterable_processing/seq_len_filter.py
Comment (FuTSy13) [inline ptls/data_load/iterable_processing/seq_len_filter.py:1]: Не очень ясна логика работы
В `IterableProcessingDataset` есть метод __iter__ в котором вызывается незаимплементированная функция  self.process.   В SeqLenFilter ни process ни __iter__ не переопределяются. Но определяется функция transform которая видимо и должна называться process

Если определена self._seq_len_col то сравнения длины последовательности с min_seq_len и max_seq_len не происходит

File: ptls/data_load/iterable_processing/take_first_trx.py
Comment (FuTSy13) [inline ptls/data_load/iterable_processing/take_first_trx.py:1]: Неправильное описание класса
Он не фильтрует последовательности, а обрезает их с конца (берет первые элементы)

File: ptls/nn/trx_encoder/trx_encoder.py
Comment (FuTSy13) [inline ptls/nn/trx_encoder/trx_encoder.py:1]: Изменение параметра по умолчанию use_batch_norm может негативно повлиять на воспроизводимость старых экспериментов

File: ptls/preprocessing/__init__.py
Comment (FuTSy13) [inline ptls/preprocessing/__init__.py:1]: может нарушиться обратная совместимость

File: ptls/preprocessing/base/__init__.py
Comment (FuTSy13) [inline ptls/preprocessing/base/__init__.py:1]: может нарушиться обратная совместимость

**Comment by cyber-felix** [comment]
Хотелось бы везде иметь docstrings и typing (возвращаемый тип в том числе)

File: ptls/data_load/datasets/memory_dataset.py [hunk 1]
```diff
@@ -28,14 +57,11 @@ def __getitem__(self, item):
 class MemoryIterableDataset(torch.utils.data.IterableDataset):
```
Comment (cyber-felix) [inline ptls/data_load/datasets/memory_dataset.py]: Может стоит указать это как абстрактный класс с абстрактным методом?

File: ptls/data_load/datasets/parquet_file_scan.py [hunk 2]
```diff
@@ -63,40 +64,45 @@ def train_valid_split(
...
+                                            valid_rate=0.3,
+                                            return_part='valid',
+                                            shuffle_seed=123)
+            for i in train_files:
+                assert i not in valid_files
+            for i in valid_files:
+                assert i not in train_files
 
     """
     assert return_part in ('train', 'valid')
```
Comment (cyber-felix) [inline ptls/data_load/datasets/parquet_file_scan.py]: Вместо assert rise ValueError(msg)

File: ptls/data_load/iterable_processing/feature_filter.py [hunk 3]
```diff
@@ -1,38 +1,49 @@
...
+
+    Args:
+        keep_feature_names: feature name for keep
+        drop_feature_names: feature name for drop
+        drop_non_iterable: drop non-iterable features
+
+    """
+
+    def __init__(self,
+                 keep_feature_names: Union[str, list] = (),
```
Comment (cyber-felix) [inline ptls/data_load/iterable_processing/feature_filter.py]: Принимается тип str или list по дефолту tuple. По итогу str добаввляется в list. Может стоит оставить Optional[Set[str]], и переписать часть кода

File: ptls/nn/trx_encoder/trx_encoder.py
Comment (technocreep) [reply ptls/nn/trx_encoder/trx_encoder.py:1]: можно, пожалуйста, чуть подробнее об этом? Потому что в целом рефакторинг негативно влияет на воспроизводимость старых экспериментов

С другой стороны `use_batch_norm` – это часть конфигурации эксперимента, в котором фигурирует `TrxEncoder`, а значит для воспроизводимости старых экспериментов достаточно поправить yaml-файлы

File: ptls/preprocessing/base/__init__.py
Comment (technocreep) [reply ptls/preprocessing/base/__init__.py:1]: здесь и выше прошу уточнить, совместимость чего с чем нарушается

File: ptls/data_load/datasets/augmentation_dataset.py
Comment (ismonnar) [reply ptls/data_load/datasets/augmentation_dataset.py:1]: Заменено

File: ptls/data_load/feature_dict.py
Comment (ismonnar) [reply ptls/data_load/feature_dict.py:1]: Добавлена обратная совместимость

File: ptls/data_load/iterable_processing/take_first_trx.py
Comment (ismonnar) [reply ptls/data_load/iterable_processing/take_first_trx.py:1]: Описание класса исправлено

File: ptls/data_load/iterable_processing/feature_filter.py (hunk 3 above)
Comment (ismonnar) [reply ptls/data_load/iterable_processing/feature_filter.py]: Оставлен Optional[Set[str]], часть кода переписана

File: ptls/data_load/datasets/parquet_file_scan.py (hunk 2 above)
Comment (ismonnar) [reply ptls/data_load/datasets/parquet_file_scan.py]: Заменено

File: ptls/data_load/datasets/memory_dataset.py (hunk 1 above)
Comment (ismonnar) [reply ptls/data_load/datasets/memory_dataset.py]: Переработано в абстрактный класс с абстрактным методом

File: ptls/data_load/iterable_processing/seq_len_filter.py
Comment (v1docq) [reply ptls/data_load/iterable_processing/seq_len_filter.py:1]: "В IterableProcessingDataset есть метод iter в котором вызывается незаимплементированная функция  self.process. В SeqLenFilter ни process ни iter не переопределяются. Но определяется функция transform которая видимо и должна называться process" 
Все верно, название функции выбрано для повышения читаемости кода и совместимости с другими библиотеками (например SkLearn). В таких решения уже определена иерархия классов, их методы и правила по которым они наследуются друг от друга. Т.К, данный класс выполняет "преобразование" входных данных, обрезая последовательности в зависимости от заданных условий, и не требует предварительного обучения "модели фильтрации", то уместным для него является определение метода transform, без метода fit. Поскольку в целях совместимости, на данном этапе разработки, оставлено наследование от IterableDataset, но при этом сам этот класс не является объектом который реализует логику итерируемого объекта (скорее наоброт - это "функция" которая применяется к итерируемому объекту), то реализовать метод iter нецелесообразно

Comment (v1docq) [reply ptls/data_load/iterable_processing/seq_len_filter.py:1]: "Если определена self._seq_len_col то сравнения длины последовательности с min_seq_len и max_seq_len не происходит"

Действительно так - здесь подразумевалась логики что если длина задана пользователем, то она уже является "валидной" и не требует проверки и сравнения с макс и мин длинами. В целом можно конечно добавить еще 1 уровень проверки для входного значения

Comment (FuTSy13) [reply ptls/data_load/iterable_processing/seq_len_filter.py:1]: > "то уместным для него является определение метода transform"

Я не нашел в коде где вызывается transform, а в остальном все понятно, со всем согласен

File: ptls/nn/trx_encoder/trx_encoder.py
Comment (FuTSy13) [reply ptls/nn/trx_encoder/trx_encoder.py:1]: Да, понятно, что после рефакторинга воспроизводимость может быть утрачена, но в таких некритичных моментах как значения аргументов по умолчанию, предлагаю оставлять как было

В старых конфигах просто не будет указан этот параметр (так как параметр по умолчанию), и, если он был важен, задебажить такое обычно очень сложно

File: ptls/preprocessing/base/__init__.py
Comment (FuTSy13) [reply ptls/preprocessing/base/__init__.py:1]:
```
from ptls.preprocessing.base import ColTransformer
```
Если где-то такое было, то код перестанет работать
Но если так и задумывалось, то ок

File: ptls/preprocessing/__init__.py
Comment (FuTSy13) [reply ptls/preprocessing/__init__.py:1]:
```
from ptls.preprocessing  import PandasDataPreprocessor
```
Если где-то такое было, то код перестанет работать
Но если так и задумывалось, то ок

File: ptls/data_load/datasets/parquet_dataset.py [hunk 4]
```diff
@@ -272,6 +253,45 @@ def __iter__(self):
...
+
+    def get_records():
+        for rb in p_table.to_batches():
+            col_arrays = [rb.column(i) for i, _ in enumerate(col_indexes)]
+            col_arrays = [a.to_numpy(zero_copy_only=False) for a in col_arrays]  # Keep as NumPy arrays
+
+            # Pre-allocate dictionary
+            records = [{} for _ in range(len(col_arrays[0]))]
+
+            if 'trans_time' in col_indexes:
```
Comment (ivkireev86) [inline ptls/data_load/datasets/parquet_dataset.py:269]: 1. `trans_time` - это старое название поля с датой. Корректное - `event_time`. Так что этот код скорее обычно не работает.
2. Не думаю, что добавлять поля сразу после чтения файла - хорошая идея. Как вариант, делать это в i_filter, или вообще в trx_encoder.

Так что предлагаю этот блок здесь убрать, и завести для работы со временем новые i_filter

File: .github/workflows/unit_test.yml [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install pipenv
+        pipenv sync --dev
+        pip install pytest pytest-cov
+
+    - name: Upgrade pandas-dask
+      run: |
+        pip install --upgrade pandas "dask[complete]"
```
Comment (ivkireev86) [inline .github/workflows/unit_test.yml]: Почему не в pipfile?

File: ptls/data_load/utils.py [hunk 6]
```diff
@@ -66,4 +71,34 @@ def collate_target(x, num=1):
...
+
+
+def collate_multimodal_feature_dict(batch):
+    res = {}
+    for source, source_batch in batch.items():
+        res[source] = collate_feature_dict(source_batch)
+    return res
+    
+    
+def get_dict_class_labels(batch):
```
Comment (ivkireev86) [inline ptls/data_load/utils.py:136]: В название функции просится слово `multimodal`

File: ptls/data_load/iterable_processing/iterable_shuffle.py [hunk 7]
```diff
@@ -1,19 +1,19 @@
-import logging
 from itertools import islice
 import numpy as np
 from ptls.data_load.iterable_processing_dataset import IterableProcessingDataset
 
 
-logger = logging.getLogger(__name__)
-
-
 class IterableShuffle(IterableProcessingDataset):
```
Comment (ivkireev86) [inline ptls/data_load/iterable_processing/iterable_shuffle.py:6]: Кажется,  в торче что-то есть на замену этому классу: `ShuffleIterDataPipe`
Но там через datapipe. Вероятно можно небольшими усилиями перейти на него

File: ptls/data_load/iterable_processing/seq_len_filter.py [hunk 8]
```diff
@@ -1,48 +1,65 @@
...
-        return len(rec[self.get_sequence_col(rec)])
+        return len(rec[self.get_sequence_col(rec)])    
+    
+    def is_len_valid(self, sequence_len: int) -> bool:
+        if self._min_seq_len is not None and sequence_len < self._min_seq_len:
+            return True
+        if self._max_seq_len is not None and sequence_len > self._max_seq_len:
+            return True
+    
+    def get_sequence_col(self, rec: dict) -> str:
```
Comment (ivkireev86) [inline ptls/data_load/iterable_processing/seq_len_filter.py]: Вроде есть в родительском классе

File: ptls/data_load/iterable_processing/seq_len_filter.py [hunk 9]
```diff
@@ -1,48 +1,65 @@
...
-
 from ptls.data_load import IterableProcessingDataset
 
-logger = logging.getLogger(__name__)
-
 
 class SeqLenFilter(IterableProcessingDataset):
-    def __init__(self, min_seq_len=None, max_seq_len=None, seq_len_col=None, sequence_col=None):
-        """
+    """
```
Comment (ivkireev86) [inline ptls/data_load/iterable_processing/seq_len_filter.py]: Вроде уехало в filtering?

File: ptls/frames/ptls_data_module.py [hunk 10]
```diff
@@ -1,27 +1,53 @@
...
+                 test_data: List[Dict] = None, 
+                 test_batch_size: int = 64,
+                 test_num_workers: int = 0,
+                 test_drop_last: bool = False,
                  ):
 
         super().__init__()
-        self.save_hyperparameters(ignore=['train_data', 'valid_data', 'test_data'])
+        self.train_batch_size = train_batch_size
+        self.train_num_workers = train_num_workers
```
Comment (ivkireev86) [inline ptls/frames/ptls_data_module.py:41]: Зачем сохранять вручную, если есть `save_hyperparameters`?

File: ptls/nn/seq_encoder/custom_encoder.py [hunk 11]
```diff
@@ -37,24 +37,24 @@ class Encoder(AbsSeqEncoder):
...
-                input_size: int,
-                intermediate_size: int = 2048, 
-                num_hidden_layers: int = 8,
-                num_attention_heads: int = 8,
-                attn_block_mode: str = 'rezero', 
-                self_attn_mode: str = 'quadratic', 
-                aggregation_mode: str = 'mean',
-                layer_norm=None,
-                is_reduce_sequence=True):
+    def __init__(self,
```
Comment (ivkireev86) [inline ptls/nn/seq_encoder/custom_encoder.py:40]: @FuTSy13 , надо бы оценить нужен этот класс сейчас или нет.

File: ptls/preprocessing/base/data_preprocessor.py [hunk 12]
```diff
@@ -1,34 +1,145 @@
 from functools import reduce
 from operator import iadd
-from typing import List
+from typing import List, Union, Callable, Dict
 
+import dask.dataframe as dd
+import pandas as pd
```
Comment (ivkireev86) [inline ptls/preprocessing/base/data_preprocessor.py:6]: Для чего pandas и dask попали в базовый препроцессор?

File: ptls/preprocessing/util.py [hunk 13]
```diff
@@ -1,24 +1,54 @@
...
+                f"Unproper `n_jobs` = {n_jobs}. "
+                f"`n_jobs` should be between ({-cpu_num}, {cpu_num}) except 0"
+            )
+        n_jobs = cpu_num + 1 + n_jobs
+    if logger:
+        logger.info(f"Number of used CPU's: {n_jobs}")
+    return n_jobs
+
+
+def dt_to_timestamp(x: pd.Series):
```
Comment (ivkireev86) [inline ptls/preprocessing/util.py:22]: Неочевидна причина переезда pandas функции в общий util.py

File: demo/CoLES-multimodal.ipynb [hunk 14]
```diff
@@ -1,4013 +0,0 @@
-{
```
Comment (ivkireev86) [inline demo/CoLES-multimodal.ipynb]: А куда делись удаленные демо? Хотелось бы оставить их.
И по демо есть потенциальный конфликт с тем что уже находится в ptls main и с https://github.com/dllllb/pytorch-lifestream/pull/172

Предлагаю конфликт решить в пользу наших изменений. От ваших взять новые файлы с примерами использования профайлера.

File: ptls/data_load/datasets/memory_dataset.py [hunk 15]
```diff
@@ -1,41 +1,67 @@
 import logging
 from typing import Iterable, List
+from abc import ABC, abstractmethod
 
+import joblib
+import pandas as pd
 import torch
-
+from pymonad.either import Either
```
Comment (ivkireev86) [inline ptls/data_load/datasets/memory_dataset.py:8]: Можете дать краткую справку по использованию pymonad? Не в коде, а скорее как руководство к дальнейшему использованию.
Сейчас встречается в правках достаточно ограниченно и не очень понятно почему. Больше нигде не требуется?

File: ptls/data_load/datasets/augmentation_dataset.py
Comment (cyber-felix) [reply ptls/data_load/datasets/augmentation_dataset.py:1]: Кажется, по этой ветке комментариев изменений нет

File: ptls/data_load/datasets/memory_dataset.py (hunk 15 above)
Comment (cyber-felix) [reply ptls/data_load/datasets/memory_dataset.py:8]: Тоже есть ощущение, что использование pymonad лишнее

File: .github/workflows/unit_test.yml (hunk 5 above)
Comment (technocreep) [reply .github/workflows/unit_test.yml]: этот момент уже не актуален, потому что экшн не существует

File: ptls/data_load/iterable_processing/seq_len_filter.py (hunk 8 above)
Comment (technocreep) [reply ptls/data_load/iterable_processing/seq_len_filter.py]: согласен

File: ptls/data_load/iterable_processing/seq_len_filter.py (hunk 9 above)
Comment (technocreep) [reply ptls/data_load/iterable_processing/seq_len_filter.py]: да, методы фильтрации, упомянутые в filtering, теперь существуют как partial классы в [filter_collection](https://github.com/ismonnar/pytorch-lifestream/blob/ptls_refactoring/ptls/data_load/iterable_processing/filter_collection.py), куда, вероятно, отправится ряд других фильтров

File: demo/CoLES-multimodal.ipynb (hunk 14 above)
Comment (technocreep) [reply demo/CoLES-multimodal.ipynb]: сделан ребейз на мэйн с решением конфликтов по демо в вашу пользу
